home: Add icons label on main navigation bar.

MritunjayTiwari14 opened the pull request:
### This PR introduces several improvements to the `HomePage` bottom navigation bar, focusing on UI enhancements for clarity, new functionality, and accessibility.

**Key Changes:**

| Dark Theme | Light Theme |
|---------|---------|
| ![1000051472](https://github.com/user-attachments/assets/dd6848fa-b72f-4f34-8c3f-b8837f6e4aa9) | ![1000051473](https://github.com/user-attachments/assets/19e987d9-67c6-4e94-a0be-5e8759f20093) |






1.  **Localizations:**
    *   Introduced corresponding localization keys (`navBarFeedLabel`, `navBarDmLabel`, `navBarMenuLabel`) in `app_en.arb`. The generated `l10n` files reflecting these additions are included.

2.  **Updated `_NavigationBarButton` Styling and Layout:**
    *   **Label Font Size:** The font size for navigation bar button labels (`_NavigationBarButton`) has been increased to `12px`. This change addresses design [feedback](https://chat.zulip.org/#narrow/channel/48-mobile/topic/Flutter.20feedback.20-.20Icons.20in.20app/with/2223128:~:text=I%20think%20we,shouldn%27t%20be%20abbreviated.) from Vlad and Alya, aiming for better readability.
    *   **Multi-line Labels:** Labels are now configured to wrap up to two lines (`maxLines: 2`) with `TextOverflow.ellipsis`. This ensures that longer localized strings can be displayed without truncation, allowing the text to naturally occupy one or two lines as needed.
    *   **Fixed Navigation Bar Height (62px):**
        *   The overall height of the bottom navigation bar is maintained at a fixed `62px` (defined by the `SizedBox` in `HomePage`).
        *   **Reasoning for Fixed Height:** While consideration was given to a fully dynamic navigation bar height that adjusts based on label content across languages, this approach was not pursued. Implementing such dynamic height adjustments introduces significant layout complexity and potential for UI jitter (where the bar changes size abruptly), which can negatively impact user experience. The current fixed height of `62px` is designed to comfortably accommodate the icon (within its `34px` `SizedBox`) and up to two lines of `12px` text. The content within each button (icon and text) is currently top-aligned.

3.  **Accessibility Improvements (`SemanticsProperties.namesRoute`):**
    *   Integrated `Semantics(namesRoute: true)` for each of the page bodies (`InboxPageBody`, `SubscriptionListPageBody`, `RecentDmConversationsPageBody`) within the `HomePage`'s `Stack`.
    *   This addresses `TODO(#535)` and enhances accessibility by allowing screen readers to announce transitions between these views (Inbox, Channels, DMs) as distinct route changes, using the `AppBar` title as the route name.

4.  **Code Formatting:**
    *   Minor code formatting adjustments, such as for the `Text()` widget in `_NavigationBarButton`, have been applied for consistency with the codebase.
    *   Updating the Label Text Color Dynamically similar to that used in iconColor using Design Variables.
   
5.  **Self Admitted Technical Dept:**
    *   Language Consistency for all language for the added labels will be fixed after receiving feedback from the Moderators about the Implementation of the solution.

These changes collectively improve the functionality, visual presentation, and accessibility of the home page navigation.

MritunjayTiwari14 commented:
@gnprice It would be extremely grateful if you could give a review to this PR for the Issue that you [stated](https://github.com/zulip/zulip-flutter/issues/1857).

gnprice commented:
Please also add a test for this change, and organize into clear and coherent commits according to the Zulip style — those are our two general requirements for a PR. See the repo's README for details, and links to further details.

Once those are met, this will be ready for a review. After another maintainer has reviewed it and you've revised to address their comments, I'll review the PR.

MritunjayTiwari14 commented:
> There is a lot of text to read in that PR description. Did you use ChatGPT or another LLM to write it? If so, we'd much rather see whatever prompt you used, instead of the LLM output — the LLM fundamentally doesn't add any information that wasn't there already, and the prompt should be a lot shorter and so less work to read.
> 
> There's also no need to repeat information that's there in the diffs, like the names of files you touched and identifiers you added there.

Although i did write the whole description in a paragraph, I use dLLM to fine tune my description into a formatted output like a pointer.

MritunjayTiwari14 commented:
@chrisbobbe This is the PR for which we discussed in the chats

For Design reference i have attached the difference.

<img width="1920" height="1080" alt="Design Provided by Vlad" src="https://github.com/user-attachments/assets/cab3462d-60b0-4905-b10c-173bf666f88c" />

I have made the commits coherent as well as minimal as request by Greg.

MritunjayTiwari14 commented:
### Before Implementing the Nav Bar Label


| Normal | Oversized System Font |
|---|---|
| ![before_l](https://github.com/user-attachments/assets/8ae7eb0d-8b52-4250-bd2c-0a42e831284a) | ![1000052413](https://github.com/user-attachments/assets/1be7e950-d8c6-49c6-a541-9991c8d917f4) |
| ![before_b](https://github.com/user-attachments/assets/414573c7-5e50-46e4-b5d6-feac8dcf0ce8) | ![1000052412](https://github.com/user-attachments/assets/56b71db9-b832-445f-a7fd-76c34ad1f160) |

<br>
<br>

### After Implementing the Nav Bar Label
| Before | After | After with Oversized System Font |
|---|---|---|
| ![before_l](https://github.com/user-attachments/assets/8ae7eb0d-8b52-4250-bd2c-0a42e831284a) | ![norm_l](https://github.com/user-attachments/assets/19e23016-d9b8-484e-ac57-7eaacf96ba7f) | ![extra_l](https://github.com/user-attachments/assets/8e0ced7b-b6b1-40ff-ae2a-e7f145dc32b7) |
| ![before_b](https://github.com/user-attachments/assets/414573c7-5e50-46e4-b5d6-feac8dcf0ce8) | ![norm_b](https://github.com/user-attachments/assets/a7eb768d-1e17-4bca-ac61-adf5d28ecb3a) | ![extra_b](https://github.com/user-attachments/assets/36e9daa2-d84c-418b-92bd-2c94acb61b17) |

<br>

_Case of Text Overflow( Upto 2 Lines )_ 

| Text Overflow Light | Text Overflow Dark |
|---|---|
| ![over_l](https://github.com/user-attachments/assets/b10fb828-0895-484f-b74b-31917d657cae) | ![over_b](https://github.com/user-attachments/assets/48ee2061-3e0b-42fb-960e-bfdd95cf7f2b) |

<br>
<br>

Thank you @chrisbobbe for your review. I feels like it better to concentrate on a single task right now (i.e the nav bar labelling).
I have made the commit history clear and coherent as well as uploaded the detailed updated screenshot comparison of the design. 

This solution also fixes the constant size of nav bar even when the system font size was increased which was a user dissatisfier, it would allow the uses to easily identify the corresponding page using the enlarged label fitted within nav bar.

chrisbobbe commented on lib/widgets/home.dart:
```diff
@@ -86,35 +86,43 @@ class _HomePageState extends State<HomePage> {
-            narrow: const CombinedFeedNarrow()))),
-      button(_HomePageTab.channels,       ZulipIcons.hash_italic),
+            narrow: const CombinedFeedNarrow())), 
+        label: zulipLocalizations.navBarFeedLabel),
```
Why not use `zulipLocalizations.combinedFeedPageTitle`, i.e. "Combined feed" in English? Was there a discussion where we decided to use "Feed" instead?

chrisbobbe commented on lib/widgets/home.dart:
```diff
@@ -86,35 +86,43 @@ class _HomePageState extends State<HomePage> {
       // TODO(#1094): Users
-      button(_HomePageTab.directMessages, ZulipIcons.two_person),
+      button(_HomePageTab.directMessages, ZulipIcons.two_person,
+        zulipLocalizations.navBarDmLabel),
```
Same as above—use `zulipLocalizations.recentDmConversationsPageTitle` here, to say "Direct messages" instead of "DMs".

(And remove the addition of `navBarDmLabel ` from your PR.)

chrisbobbe commented on lib/widgets/home.dart:
```diff
@@ -231,35 +237,50 @@ class _NavigationBarButton extends StatelessWidget {
+        child: InkWell(
           // TODO(#417): Disable splash effects for all buttons globally.
+          borderRadius: BorderRadius.all(Radius.circular(4)),
           splashFactory: NoSplash.splashFactory,
```
nit: don't separate the TODO from the line of code it is about

chrisbobbe reviewed (changes requested):
Thanks!

For the changes proposed in these 7 commits, I think it's actually better to squash them into one single commit. The desired changes aren't very complicated, and it's helpful to see them all together in a coherent whole. 🙂 This approach would also clean up the "noise" of a commit like this one:

```
nits: Remove comma after color property of Icon in home bottom navbar button
```

That kind of commit never needs to exist. If there's a problem introduced in an earlier commit in the PR, the solution is to change the earlier commit to not introduce the problem, rather than adding a new commit, which just adds work in PR review and when reading Git history.

Please also add a test for this change, as Greg pointed out above.

MritunjayTiwari14 commented on lib/widgets/home.dart:
```diff
@@ -231,35 +237,50 @@ class _NavigationBarButton extends StatelessWidget {
+                      height: 1.0,),
+                    textAlign: TextAlign.center,
+                    maxLines: 2,
+                    overflow: TextOverflow.ellipsis)),
```
The naming of Direct Message to DMs was discussed by Tim over here [#mobile > Flutter feedback - Icons in app @ 💬](https://chat.zulip.org/#narrow/channel/48-mobile/topic/Flutter.20feedback.20-.20Icons.20in.20app/near/2262208)

> <img width="500" height="200" alt="image" src="https://github.com/user-attachments/assets/cc0c8d7d-77a5-47a4-9b4f-9c45b6214d1c" />
<br>

and, the maxLines and overflow case by Greg over here [#mobile > Flutter feedback - Icons in app @ 💬](https://chat.zulip.org/#narrow/channel/48-mobile/topic/Flutter.20feedback.20-.20Icons.20in.20app/near/2223128)

> <img width="1000" height="341" alt="image" src="https://github.com/user-attachments/assets/92119151-429c-4f55-ad6e-76e00a869b8d" />


<br>
However, if the translator is willing to provide a shorter label then i think it would be best to eliminate the overflow and maxLines(just in case if the label gets too much big to disturb the main page).

MritunjayTiwari14 commented:
Thank you @chrisbobbe for the detailed feedback, the text font scaling has been handled using `textScaler` with a `maxScaleFacctor` as requested with the other changes also.

<details>
<summary>Comparison Throught Screenshots</summary>

### Maxing Android Font and Display Size:

| Before | After |
| --- | --- |
| <img width="1280" height="2856" alt="Screenshot_1761559270" src="https://github.com/user-attachments/assets/7160d345-3963-45ad-a2f4-66f3e1801418" /> | <img width="1280" height="2856" alt="Screenshot_1761559233" src="https://github.com/user-attachments/assets/93204859-751a-4308-901b-df9b9494fd99" /> |

| Before | After |
| --- | --- |
| <img width="1280" height="2856" alt="Screenshot_1761559263" src="https://github.com/user-attachments/assets/f51c885b-2770-4678-a993-6cead71a51a8" /> | <img width="1280" height="2856" alt="Screenshot_1761559248" src="https://github.com/user-attachments/assets/2702175d-a594-4f16-8bfa-36ef7686a5d0" /> |

</details>

MritunjayTiwari14 commented on lib/widgets/home.dart:
```diff
@@ -231,35 +237,50 @@ class _NavigationBarButton extends StatelessWidget {
+              mainAxisSize: MainAxisSize.min,
+              children: [
+                SizedBox(height: 34,
+                  child: Center(
```
Thank you for Noticing, that why when I saw the UI it looks overpadded.

MritunjayTiwari14 commented:
@gnprice Thank you for updating regarding the new issue filed which can be easily solve within this.

After trying for few hours, I would like to point out few things regarding the testing of TalkBack. From the trials i have done. The TalkBack was being displayed neither by Live Transcript not via any Screen Recording software which has support to record android internal audio. Hence i think its not possible to able to capture it and display it over here.

However, I have tested this on my android device and the changes works as expected.

chrisbobbe reviewed (commented):
Thanks! Comments below.

MritunjayTiwari14 commented:
Thank you @chrisbobbe for the review. Have cleared out some nits as well as given `SemanticsRole.tab` to the `Semantics` and the other necessary things for it, PTAL.

chrisbobbe commented:
Thanks! I've tested that out with VoiceOver and found a way to improve the experience. See the two commits I've pushed on top of yours; here's the list of all three:

8f1523383 home: Add label to bottom navbar icons
ca48d8897 home [nfc]: Factor bottom nav into its own widget
28ad21542 home: Adjust semantics of bottom nav bar

I'll mark this for Greg's review. If he requests changes, please pull the branch including my commits so we don't lose them, before applying your changes to the branch.

gnprice commented on lib/widgets/home.dart:
```diff
@@ -93,58 +97,89 @@ class _HomePageState extends State<HomePage> {
+      onPressed: () {
+        tabNotifier.value = tab;
+      },
+      label: label);
```
nit: keep label next to icon (here and below) — they're logically closely related, as the two pieces of visible content

gnprice commented on lib/widgets/home.dart:
```diff
@@ -93,58 +97,89 @@ class _HomePageState extends State<HomePage> {
+        identifier: HomePage.contentSemanticsIdentifier,
+        container: true,
+        explicitChildNodes: true,
+        child: Stack(
```
Hmm, interesting.

    The rest is just a best effort at following the docs, which doesn't
    make the experience obviously worse, but I didn't notice an effect
    in my testing.

@chrisbobbe Were there particular docs you found especially helpful which you can link to? I'm curious for things we can consult in the future. (No need to spend a lot of time digging now, though — if things seem to behave well, that's good enough for the code at this stage.)

chrisbobbe replied:
Just the docs for the things I used. I found them by just poking around to see if there was some structured way to describe the UI to screen-reader software:

https://api.flutter.dev/flutter/dart-ui/SemanticsRole.html#tab
https://api.flutter.dev/flutter/dart-ui/SemanticsRole.html#tabBar
https://api.flutter.dev/flutter/dart-ui/SemanticsRole.html#tabPanel
https://api.flutter.dev/flutter/semantics/SemanticsNode/controlsNodes.html
https://api.flutter.dev/flutter/semantics/SemanticsProperties/namesRoute.html

gnprice reviewed (commented):
Thanks @MritunjayTiwari14 for taking care of this, and @chrisbobbe for the previous reviews! Comments below.

gnprice commented:
Several of the comments in my previous review are about changes that happen in the first commit of the branch. To make the commits [clean and coherent](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html), that means the revisions should apply to that first commit, not to a later commit.

You'll use `git rebase -i` to do that. For background, see [our Git guide](https://zulip.readthedocs.io/en/latest/git/overview.html), and you can ask questions in the `#git help` channel.

MritunjayTiwari14 commented:
Thank you @gnprice for the suggestion. Implemented `Center` with `heightFactor` 1, which gives identical results. Also i have made changes to Chris's commits due to conflict arising while making edits to my previous commit so i have added Chris as co-author in those commits. 

Hope that the commits are now coherent and clean.

gnprice reviewed (commented):
Thanks for the revision. These changes generally all look good.

> Also i have made changes to Chris's commits due to conflict arising while making edits to my previous commit so i have added Chris as co-author in those commits.

Are you using `git rebase -i`? That should automatically preserve the authorship information as you revise the commits.

If you're not already using `git rebase -i`, it's an important skill and well worth spending some time to learn.

When you revise a commit and the only effect is resolving rebase conflicts, and/or trivial formatting changes, the authorship information should be kept the same as in the original commits. That's the case here, so please revise these to put the original authorship information back.

You can do that with a command `git commit --amend -C $other_commit_id`. See `git help commit` for docs. You'll want to use `git rebase -i` in order to get at both commits. Again, the `#git help` channel is a good place to ask any questions about how to use `git rebase -i` or any other aspect of Git.

gnprice commented:
Hmm, the authorship information still doesn't match the original commits:
```
commit 83047c1a0bf97b4a9654b9163765f27d773794a6 (HEAD -> main, pr/1879)
Author: chrisbobbe <cbobbe@zulip.com>
Date:   Sat Nov 8 14:20:30 2025 +0530
```

If you use `git commit --amend -C`, it will reliably get that correct. Please do use that command. If you need help:

> See `git help commit` for docs. […] Again, the `#git help` channel is a good place to ask any questions about how to use `git rebase -i` or any other aspect of Git.

MritunjayTiwari14 commented:
Thank you @gnprice for noticing the inconsistency of commit history, the original commit of chris which was as :

```
commit a3f87f9f804d20920e3fb0a87f05e2f9ec842ade
Author: Chris Bobbe <cbobbe@zulip.com>
Date:   Wed Nov 5 12:07:47 2025 -0800

    home [nfc]: Factor bottom nav into its own widget
```

has been restore as per the given command `git commit --amend -C` with the commit hash provided by chris in the comments above.

MritunjayTiwari14 commented:
Thank you @gnprice for noticing it, have changed the description as given.

gnprice commented:
Thanks! This looks good now; merging. I've fixed one nit in a commit message:

         home: Adjust semantics of bottom nav bar
     
         Fixes #1960.
    +
         Some of this has an effect I observes in my testing on iOS with

namely to put a paragraph break after the "Fixes …" line.

> I personally feels like my first commit [home: Add label to bottom navbar icons](https://github.com/zulip/zulip-flutter/pull/1879/commits/5ee186d29149551b7ae4b9b3ed47e1ae1fef6090) is partially fixing the semantic on navbar which does not seems to be coherent and would be difficult to bisect.
> 
> It think its better to keep the semantic updates to Chris's commit [home: Adjust semantics of bottom nav bar](https://github.com/zulip/zulip-flutter/pull/1879/commits/688f9a56f2987f100145998fc999e8e1d3b33514) and remove them from mine.

Yeah, I think the ideal version of this set of commits would probably separate the semantics changes from the label change differently. This version is clear enough, though. Thanks for all your work on it!
